### PR TITLE
KeyMapping and expanded input handling

### PR DESCRIPTION
**Pull request by JamesHansIV**
* first iteration of custom cursor implementation. Will be show the user what tool they are currently using (pencil for draw, eraser for erase, etc.) for a better user experience and visibility.
* Created KeyMap class and implementation. Allows specific EditorActions (a new enum class) to be mapped to specific keys (or multiple keys). Eliminates hard coding inputs and allows for custom key binding on both user and dev side. Could potentially be implemented in the Game as well.
* Updated InputChecker to include track the keys that were pressed in the previous frame. This is used for latch type inputs (key must be released before the second input is registered).
* Implemented updated input handling and keymap into the editor
* updated the todo list
* Included the first iterations (light and dark modes) of the pencil (draw) cursor

### CODE REVIEW

File: todo.txt [hunk 1]
```diff
@@ -7,6 +9,30 @@ x managing health should be done by the health class
 - add enemy types
 x enemies are able to slip through walls somehow
 
+editor stuff
+- resizable main window
+x switch edit mode from keybinds (decouple edit mode from gui, so we can change the gui to new layout)
+~ Visual indicator for edit mode
+    - Text
+    ~ Cursor
+- Erase tool should not have to collide with with the texture game object to delete the texture.
```
Comment (lee892) [inline todo.txt:18]: I can't remember if all objects within the same tile and layer are deleted at once with the erase tool, but they should if they don't

File: src/Engine/KeyMap/KeyMap.cpp [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+Bind KeyMap::CreateBind(EditorAction action, InputType input_type, std::vector<SDL_KeyCode> keys) {
+    Bind bind = {action, input_type, keys};
+    return bind;
+}
+
+
+KeyMap::KeyMap() {
+    m_Default_Bindings_Map = { 
+        {EditorAction::EXIT_CURRENT_TOOL, CreateBind(EditorAction::ENTER_DRAW_TOOL, InputType::LATCH, std::vector<SDL_KeyCode>{SDLK_ESCAPE})},
```
Comment (lee892) [inline src/Engine/KeyMap/KeyMap.cpp:31]: Should EXIT_CURRENT_TOOL be mapped to ENTER_DRAW_TOOL?

File: src/Engine/Editor/Editor.cpp [hunk 3]
```diff
@@ -220,6 +221,25 @@ Editor::Editor() {
...
     m_Layers.emplace_back();
+
+    // create keymap
+    m_KeyMap = new KeyMap();
+
+    std::cout << "Keymap createed\n";
+
+    // add cursor textures to renderer
+    std::cout << "Adding draw texture... ";
+    Renderer::GetInstance()->AddTexture(
```
[the remark this answers is not in the document]
Comment (JamesHansIV) [reply src/Engine/Editor/Editor.cpp:232]: Textures are being added a separate editor_textures.xml file. Will be in next PR.

File: todo.txt [hunk 4]
```diff
@@ -7,6 +9,30 @@ x managing health should be done by the health class
...
+- create multiselect tool 
+  - select
+  - copy
+  - paste
+  - delete
+- save scenes (to be loaded into game application)
+.
+.
+.
+(would be cool)
```
[the remark this answers is not in the document]
Comment (JamesHansIV) [reply todo.txt:33]: 🫠 damn...

File: src/Engine/KeyMap/KeyMap.cpp (hunk 2 above)
Comment (JamesHansIV) [reply src/Engine/KeyMap/KeyMap.cpp:31]: No. Whoops. I'll get that fixed.

File: todo.txt (hunk 1 above)
Comment (JamesHansIV) [reply todo.txt:18]: Current behavior is that it the erase tool doesn't delete all object's on the same tile and layer. I'll fix that with the next PR.

File: src/Engine/Input/InputChecker.cpp [hunk 5]
```diff
@@ -46,4 +53,16 @@ void InputChecker::SetMouseWheelDirection(int direction) {
 
 int InputChecker::GetMouseWheelDirection() {
     return mouse_wheel_direction;
+}
+
+void InputChecker::SetPrevFrameKeys() {
```
[the remark this answers is not in the document]
Comment (JamesHansIV) [reply src/Engine/Input/InputChecker.cpp:58]: I'm not noticing any performance issues. If we run into issues down the line I'll rewrite it.

File: src/Engine/Editor/Editor.cpp [hunk 6]
```diff
@@ -457,7 +480,7 @@ void Editor::ShowFileManager() {
     if (ImGui::BeginTabItem("File")) {
         ImGui::Text("Current room: %s", m_CurrentRoomID.c_str());
 
-        if (ImGui::Button("Save Project", ImVec2(150, 20))) {
+        if (ImGui::Button("Save SDFSDFSDProject", ImVec2(150, 20))) {
```
[the remark this answers is not in the document]
Comment (JamesHansIV) [reply src/Engine/Editor/Editor.cpp:483]: Tomfoolery

File: src/Engine/Editor/Editor.cpp [hunk 7]
```diff
@@ -1268,6 +1315,11 @@ void Editor::Events() {
     }
 }
 
+void Editor::CheckForEditorModeChangingInputs(EditorAction editor_action, EditMode edit_mode) {
```
[the remark this answers is not in the document]
Comment (JamesHansIV) [reply src/Engine/Editor/Editor.cpp:1318]: This is a function to handle inputs that cause edit mode to change (tool selection / deselection via hotkey).

The thought here was that for every input that changes the edit mode there is going to be an if statement, followed by a ternary to handle the "toggle" behavior, then a statement setting the cursor's type. 

The logic behind tool selection via keyboard input is going to the same (toggling mode if needed & updating cursor) for each tool we implement. Tool behavior is implemented in Editor::OnMouseClicked, so I don't really see a point in writing out all that repeated code for each EditorAction that causes a EditMode change. 

I do think that the that function name is horrible and can cause confusion about what its really doing. I'll update that to something more clear and we can keep discussing whether or not the function needs to exist at all.

File: src/Engine/Editor/Editor.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (lee892) [reply src/Engine/Editor/Editor.cpp:473]: I love it

File: src/Engine/Editor/Editor.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (lee892) [reply src/Engine/Editor/Editor.cpp]: Ah I was just confused as to why a function contained a single if statement, but I guess if it's going to be used multiple times it should be fine

**Pull request merged**